### PR TITLE
Handling of files for which no importer (or too many) is present does not fail with an exception

### PR DESCRIPTION
**Pull request by poldi2015**
- [x] Implement files for ZIP File Importer
- [x] Improve ZIP File importer
- [x] Implement tests

### CODE REVIEW

File: core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java [hunk 1]
```diff
@@ -131,12 +132,18 @@ public MultiFileImporter importRecursiveDir(final Path dir, final String glob)
...
-                + "' for file '" + file + "'");
+        final Optional<ImporterFactory> importerFactory = this.factoryLoader.getImporterFactory(file);
+        final Optional<Importer> importer = importerFactory.isPresent() ?
+                Optional.of(importerFactory.get().createImporter(file, builder)) :
+                Optional.empty();
+
+        LOG.fine(() -> ( importer.isPresent() ?
+                            "Created importer of type '" + importer.getClass().getSimpleName() :
+                            "No import" )
+                       + "' for file '" + file + "'");
```
Comment (kaklakariada) [inline core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java:145]: 👍

File: core/src/test/java/org/itsallcode/openfasttrace/core/importer/TestImporterFactoryLoader.java [hunk 2]
```diff
@@ -78,10 +80,10 @@ void beforeEach()
     }
 
     @Test
-    void testNoFactoryRegisteredThrowsException()
+    void testNoFactoryRegisteredReturnsNoImporter()
     {
         simulateFactories();
-        assertThrows(ImporterException.class, () -> this.loader.getImporterFactory(this.file));
+        assertEquals(Optional.empty(),this.loader.getImporterFactory(this.file));
```
Comment (kaklakariada) [inline core/src/test/java/org/itsallcode/openfasttrace/core/importer/TestImporterFactoryLoader.java:86]:
```suggestion
        assertTrue(this.loader.getImporterFactory(this.file).isEmpty());
```
Besser❓

**Review by kaklakariada — APPROVED** [review]

File: core/src/test/java/org/itsallcode/openfasttrace/core/importer/TestImporterFactoryLoader.java (hunk 2 above)
Comment (poldi2015) [reply core/src/test/java/org/itsallcode/openfasttrace/core/importer/TestImporterFactoryLoader.java]: At least also good :-). Worked in proposed changes.

**Comment by kaklakariada** [comment]
Bitte noch Requirements und Design aktualisieren :)

**Review by kaklakariada — APPROVED** [review]

**Comment by kaklakariada** [comment]
Build for PR from fork failed because of missing SONAR_TOKEN. We still can merge this.
@redcatbear Please also review.

**Comment by poldi2015** [comment]
Can we merge this pull request to develop or is something still open? Should @redcatbear  also do a review?

**Comment by kaklakariada** [comment]
Yes, he should also review.

File: api/src/main/java/org/itsallcode/openfasttrace/api/importer/MultiFileImporter.java [hunk 3]
```diff
@@ -22,6 +22,8 @@
  * #L%
  */
 
+
+
```
Comment (redcatbear) [inline api/src/main/java/org/itsallcode/openfasttrace/api/importer/MultiFileImporter.java:26]: Unnecessary change.

File: core/src/main/java/org/itsallcode/openfasttrace/core/importer/ImporterFactoryLoader.java [hunk 4]
```diff
@@ -63,18 +64,19 @@ public ImporterFactoryLoader(
      * @throws ImporterException
      *             when no or more than one {@link ImporterFactory} is found.
      */
-    public ImporterFactory getImporterFactory(final InputFile file)
+    public Optional<ImporterFactory> getImporterFactory(final InputFile file)
```
Comment (redcatbear) [inline core/src/main/java/org/itsallcode/openfasttrace/core/importer/ImporterFactoryLoader.java:67]: Please split into check method and getter method instead of using `Optional`. Use getter only if check comes back green.

File: core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java [hunk 5]
```diff
@@ -131,12 +132,18 @@ public MultiFileImporter importRecursiveDir(final Path dir, final String glob)
...
-        LOG.fine(() -> "Created importer of type '" + importer.getClass().getSimpleName()
-                + "' for file '" + file + "'");
+        final Optional<ImporterFactory> importerFactory = this.factoryLoader.getImporterFactory(file);
+        final Optional<Importer> importer = importerFactory.isPresent() ?
+                Optional.of(importerFactory.get().createImporter(file, builder)) :
+                Optional.empty();
+
+        LOG.fine(() -> ( importer.isPresent() ?
+                            "Created importer of type '" + importer.getClass().getSimpleName() :
+                            "No import" )
```
Comment (redcatbear) [inline core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java:144]:
```suggestion
        LOG.fine(() -> (importer.isPresent() ?
                            "Created importer of type '" + importer.getClass().getSimpleName() :
                            "No import")
```
Please use auto-formatting. that should also make sure that all spaces are uniform.

**Review by redcatbear — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/itsallcode/openfasttrace/api/importer/MultiFileImporter.java (hunk 3 above)
Comment (poldi2015) [reply api/src/main/java/org/itsallcode/openfasttrace/api/importer/MultiFileImporter.java:26]: DONE

File: core/src/main/java/org/itsallcode/openfasttrace/core/importer/ImporterFactoryLoader.java (hunk 4 above)
Comment (poldi2015) [reply core/src/main/java/org/itsallcode/openfasttrace/core/importer/ImporterFactoryLoader.java:67]: This would mean that the ImporterFactory needs to be searched twice: once by the check method to figure out that there is  a fitting ImporterFactory and second by the get method. Th search calls ImporterFactory.supportsFile() on each factory when then figures out by examine the given file if it can load the file.

What is the problem with `Optional`. As Java has no `Either` construct, isn't Optional not the Java approach for such a construct?

File: core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java (hunk 5 above)
Comment (poldi2015) [reply core/src/main/java/org/itsallcode/openfasttrace/core/importer/MultiFileImporterImpl.java:144]: DONE

**Review by redcatbear — APPROVED** [review]

**Comment by redcatbear** [comment]
Build failing.

**Comment by kaklakariada** [comment]
Build fails during sonar analysis with message `Caused by: org.sonar.api.utils.MessageException: You're not authorized to run analysis. Please contact the project administrator.`

Root cause is, that pr from forks don't have access to the sonar token. I fixed this issue on `develop` by skipping sonar in this case.

I will merge the pr as is. We can fix potential sonar findings later.
